File: python/2303.calculate-amount-paid-in-taxes.py
#
# @lc app=leetcode id=2303 lang=python3
#
# [2303] Calculate Amount Paid in Taxes
#

# @lc code=start
from typing import List


class Solution:
    def calculateTax(self, brackets: List[List[int]], income: int) -> float:
        prev = 0
        tax = 0.0
        for l1 in brackets:
            if income < l1[0] and income < prev:
                break
            tax += (min(income, l1[0]) - prev) * (0.01 * l1[1])
            prev = l1[0]
        return tax

# obj = Solution()
# print(obj.calculateTax([[3,50],[7,10],[12,25]], 10))
# print(obj.calculateTax([[1,0],[4,25],[5,50]], 2))
        
# @lc code=end

File: 2303.calculate-amount-paid-in-taxes.java
/*
 * @lc app=leetcode id=2303 lang=java
 *
 * [2303] Calculate Amount Paid in Taxes
 */

// @lc code=start
class Solution {
    public double calculateTax(int[][] brackets, int income) {
        DecimalFormat df = new DecimalFormat("#.#####");
        double tax = 0;
        for (int i = 0; i < brackets.length; i++) {
            double taxableIncome = 0;
            if (brackets[i][0] - income < 0) {
                if (i == 0) {
                    taxableIncome = brackets[i][0];
                } else {
                    taxableIncome = brackets[i][0] - brackets[i - 1][0];
                }

            } else {
                if (i == 0) {
                    if(income > 0) { 
                        if(income == brackets[i][0] || brackets[i][0] > income)
                            taxableIncome = income;
                        else
                            // taxableIncome = Math.abs(income - brackets[i][0]);
                            taxableIncome = brackets[i][0];
                    } else {
                        taxableIncome = income;
                    }
                } else {
                    taxableIncome = income - brackets[i - 1][0];
                }

            }
            if (taxableIncome < 0)
                continue;
            double percent = ((double) brackets[i][1] / 100);
            tax += taxableIncome * percent;

        }
        return tax;
    }
}
// @lc code=end
